### PR TITLE
cdc/scheduler_v2: implement the basic logic for the new schedule dispatcher

### PR DESCRIPTION
**Pull request by liuzix**
**Notes for Reviewers: More than half of the LOCs are tests. Non-test code is about 500 lines.**

### What problem does this PR solve? 
- The peer-messaging based scheduler. For a high-level design doc, refer to #3363 

### What is changed and how it works?
- Implemented a peer-messaging based scheduler.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/scheduler/scheduler.go [hunk 1]
```diff
@@ -0,0 +1,626 @@
...
+// This interface is akin to a Virtual Method Table, in that the invoker of
+// BaseScheduleDispatcher should provide an implementation of ScheduleDispatcherCommunicator
+// to supply BaseScheduleDispatcher some methods to specify its behavior.
+type ScheduleDispatcherCommunicator interface {
+	// DispatchTable should send a dispatch command to the Processor.
+	DispatchTable(ctx context.Context,
+		changeFeedID model.ChangeFeedID,
+		tableID model.TableID,
+		captureID model.CaptureID,
+		boundaryTs model.Ts, // Deprecated. Kept here for API compatibility.
```
Comment (3AceShowHand) [inline cdc/scheduler/scheduler.go:59]: can you be more clear about this should keep API compatibility with which one?

File: cdc/scheduler/scheduler.go [hunk 2]
```diff
@@ -0,0 +1,626 @@
...
+	moveTableJobQueue []*moveTableJob
+	moveTableTarget   map[model.TableID]model.CaptureID // caches the user specified targets to move tables to.
+
+	lastTickCaptureCount int
+	needRebalance        bool
+
+	// read only fields
+	changeFeedID model.ChangeFeedID
+	communicator ScheduleDispatcherCommunicator
+	logger       *zap.Logger
```
Comment (3AceShowHand) [inline cdc/scheduler/scheduler.go:94]: why not use the global logger?

File: cdc/scheduler/scheduler.go [hunk 3]
```diff
@@ -0,0 +1,626 @@
...
+}
+
+// NewBaseScheduleDispatcher creates a new BaseScheduleDispatcher.
+func NewBaseScheduleDispatcher(
+	changeFeedID model.ChangeFeedID,
+	communicator ScheduleDispatcherCommunicator,
+	checkpointTs model.Ts,
+) *BaseScheduleDispatcher {
+	logger := log.L().With(zap.String("changefeed-id", changeFeedID))
```
Comment (3AceShowHand) [inline cdc/scheduler/scheduler.go:103]: will this use the global logger for logging?

File: cdc/scheduler/scheduler.go [hunk 4]
```diff
@@ -0,0 +1,659 @@
...
+	status, ok := s.captureStatus[captureID]
+	if !ok {
+		s.logger.Info("received checkpoint from non-existing capture, ignore",
+			zap.String("capture-id", captureID),
+			zap.Uint64("checkpoint-ts", checkpointTs),
+			zap.Uint64("resolved-ts", resolvedTs))
+		return
+	}
+
+	if status.SyncStatus != captureSyncFinished {
```
[the remark this answers is not in the document]
Comment (liuzix) [reply cdc/scheduler/scheduler.go:649]: A capture that has not finished sync can still send checkpoints.

File: cdc/scheduler/scheduler_dispatcher.go [hunk 5]
```diff
@@ -62,3 +69,550 @@ type ScheduleDispatcherCommunicator interface {
...
+	// captureSyncFinished indicates that the capture has been fully initialized and is ready to
+	// accept `DispatchTable` messages.
+	captureSyncFinished
+)
+
+// Tick implements the interface ScheduleDispatcher.
+func (s *BaseScheduleDispatcher) Tick(
+	ctx context.Context,
+	checkpointTs model.Ts,
+	currentTables []model.TableID,
```
Comment (amyangfei) [inline cdc/scheduler/scheduler_dispatcher.go:152]: What does `currentTables` mean and how is it maintained?

Comment (liuzix) [reply cdc/scheduler/scheduler_dispatcher.go]: I added some comments. It is maintained by `schemaWrap4Owner` defined in `cdc/owner/schema.go`.

File: cdc/scheduler/schedule_dispatcher.go [hunk 6]
```diff
@@ -0,0 +1,626 @@
...
+	// Makes sure that captures have all been synchronized before proceeding.
+	done, err := s.syncCaptures(ctx)
+	if err != nil {
+		return CheckpointCannotProceed, CheckpointCannotProceed, errors.Trace(err)
+	}
+	if !done {
+		// Returns early if not all captures have synced their states with us.
+		// We need to know all captures' status in order to proceed.
+		// This is crucial for ensuring that no table is double-scheduled.
+		return CheckpointCannotProceed, CheckpointCannotProceed, nil
```
Comment (3AceShowHand) [inline cdc/scheduler/schedule_dispatcher.go:182]: do we have to add a log here to indicate that not all captured synced yet?

File: cdc/scheduler/schedule_dispatcher.go [hunk 7]
```diff
@@ -0,0 +1,626 @@
...
+	// these operations.
+	toAdd, toRemove := s.findDiffTables(shouldReplicateTableSet)
+
+	for _, tableID := range toAdd {
+		ok, err := s.addTable(ctx, tableID)
+		if err != nil {
+			return CheckpointCannotProceed, CheckpointCannotProceed, errors.Trace(err)
+		}
+		if !ok {
+			return CheckpointCannotProceed, CheckpointCannotProceed, nil
```
Comment (3AceShowHand) [inline cdc/scheduler/schedule_dispatcher.go:205]: ditto

File: cdc/scheduler/schedule_dispatcher.go [hunk 8]
```diff
@@ -0,0 +1,626 @@
...
+			// another operation is in progress
+			continue
+		}
+
+		ok, err := s.removeTable(ctx, tableID)
+		if err != nil {
+			return CheckpointCannotProceed, CheckpointCannotProceed, errors.Trace(err)
+		}
+		if !ok {
+			return CheckpointCannotProceed, CheckpointCannotProceed, nil
```
Comment (3AceShowHand) [inline cdc/scheduler/schedule_dispatcher.go:224]: ditto

File: cdc/scheduler/schedule_dispatcher.go [hunk 9]
```diff
@@ -0,0 +1,626 @@
...
+		return CheckpointCannotProceed, CheckpointCannotProceed, nil
+	}
+
+	// handleMoveTableJobs tries to execute user-specified manual move table jobs.
+	ok, err := s.handleMoveTableJobs(ctx)
+	if err != nil {
+		return CheckpointCannotProceed, CheckpointCannotProceed, errors.Trace(err)
+	}
+	if !ok {
+		return CheckpointCannotProceed, CheckpointCannotProceed, nil
```
Comment (3AceShowHand) [inline cdc/scheduler/schedule_dispatcher.go:243]: ditto

File: cdc/scheduler/schedule_dispatcher.go [hunk 10]
```diff
@@ -0,0 +1,626 @@
...
+		s.needRebalance = true
+	}
+	s.lastTickCaptureCount = len(captures)
+
+	if s.needRebalance {
+		ok, err := s.rebalance(ctx)
+		if err != nil {
+			return CheckpointCannotProceed, CheckpointCannotProceed, errors.Trace(err)
+		}
+		if ok {
```
Comment (3AceShowHand) [inline cdc/scheduler/schedule_dispatcher.go:264]: how about check `!ok` first, and remove the `else` branch.

File: cdc/scheduler/schedule_dispatcher.go (hunk 6 above)
Comment (liuzix) [reply cdc/scheduler/schedule_dispatcher.go:182]: There can be too many log messages, especially during a delta scan. There can easily be several messages per second per changefeed.

**Review by 3AceShowHand — APPROVED** [review]

**Review by amyangfei — APPROVED** [review]

File: cdc/scheduler/schedule_dispatcher.go [hunk 11]
```diff
@@ -0,0 +1,624 @@
...
+	// captureSyncFinished indicates that the capture has been fully initialized and is ready to
+	// accept `DispatchTable` messages.
+	captureSyncFinished
+)
+
+// Tick implements the interface ScheduleDispatcher.
+func (s *BaseScheduleDispatcher) Tick(
+	ctx context.Context,
+	checkpointTs model.Ts,
+	// currentTables are tables that SHOULD be running given the current checkpoint-ts.
```
Comment (zhaoxinyu) [inline cdc/scheduler/schedule_dispatcher.go:152]: In this line, 'SHOUD be running given the current checkpoint-ts' is a little bit hard to understand

**Pull request merged**